spec/Section 3 -- Type System.md

```diff
@@ -1340,14 +1357,281 @@ Union type extensions have the potential to be invalid if incorrectly defined.
+`$var`                   | `{ var: { a: "abc" } }` | `{ a: "abc" }`
+`{ a: "abc", b: null }`  | `{}`                    | Error: Exactly one key must be specified
+`{ b: $var }`            | `{ var: null }`         | Error: {b} must be non-null.
+`{ b: 123, c: "xyz" }`   | `{}`                    | Error: Unexpected member {c}
+
+
+
+**Type Validation**
+
+Tagged types have the potential to be invalid if incorrectly defined. This set
+of rules must be adhered to by every Tagged type in a GraphQL schema.
+
+1. A Tagged type must define one or more members.
+2. For each member of a Tagged type:
+   1. The member must have a unique name within that Tagged type;
+      no two members may share the same name.
+   2. The member must not have a name which begins with the
+      characters {"__"} (two underscores).
+3. {IsOutputType(taggedType)} and {IsInputType(taggedType)} cannot both be {false}.
+
+### Member Deprecation
+
+Members in a Tagged type may be marked as deprecated as deemed necessary by
+the application. It is still legal to query for these members on outputs or
+supply these members in inputs (to ensure existing clients are not broken by
+the change), but the members should be appropriately treated in documentation
+and tooling.
+
+When using the type system definition language, `@deprecated` directives are
+used to indicate that a field is deprecated:
+
+```graphql example
+tagged ExampleType {
+  oldField: String @deprecated
+}
+```
+
+Note: a deprecated member of a tagged type may legitimately be marked as
+non-nullable since this member need not be supplied/returned.
+
+### Tagged Extensions
+
+TaggedTypeExtension :
+  - extend tagged Name Directives[Const]? TaggedMembersDefinition
+  - extend tagged Name Directives[Const]
+
+Tagged type extensions are used to represent a Tagged type which has been
+extended from some original Tagged type. For example, this might be used to
+represent a GraphQL service which is itself an extension of another GraphQL
+service.
+
+In this example, an additional option `startsWith` is added to a `StringFilter`
+type:
+
+```graphql example
+extend tagged StringFilter {
+  startsWith: String!
+}
+```
+
+Tagged type extensions may choose not to add additional members, instead only
+adding directives.
+
+In this example, a directive is added to a `StringFilter` type without adding
+members:
+
+```graphql example
+extend tagged StringFilter @addedDirective
+```
+
+**Type Validation**
+
+Tagged type extensions have the potential to be invalid if incorrectly defined.
+
+1. The named type must already be defined and must be a Tagged type.
+2. The members of a Tagged type extension must have unique names; no two members
+   may share the same name.
+3. The members of a Tagged type extension must not have names beginning with
+   {"__"}.
+4. Any members of a Tagged type extension must not be already defined on the
+   original Tagged type.
+5. Any non-repeatable directives provided must not already apply to the
```

Comment:
You might want to also repeat the rule that after adding new members, IsInputType and IsOutputType can't both be false

Reply:
:+1: